### PR TITLE
[InstCombine] Fold `(mul (div exact X, C0), C1)` -> `(div exact X, C0/C1)`

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[InstCombine] Add tests for folding `(mul (div exact X, C0), C1)`; NFC**
- **[InstCombine] Fold `(mul (div exact X, C0), C1)` -> `(div exact X, C0/C1)`**

We can do this if `C0 % C1 == 0` and if we avoid UB in the signed
case.

Proofs: https://alive2.llvm.org/ce/z/HHWHDg

This is essentially the same as #96898.
NB: No need to handle shifts. We already have folds for them and mul+shr/shl+div will never fold (assuming any pow2 constants for mul/div are converted to shifts).

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 1]
```diff
@@ -365,6 +365,28 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
       return BinaryOperator::CreateMul(NegOp0, X);
   }
 
+  if (Op0->hasOneUse()) {
+    // (mul (div exact X, C0), C1)
+    //    -> (div exact C0 / C1)
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:370]:
```suggestion
    //    -> (div exact X, C0 / C1)
```

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 2]
```diff
@@ -365,6 +365,28 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
...
+    // iff C0 % C1 == 0 and C0 / C1 doesn't create UB.
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:371]:
```suggestion
    // iff C0 % C1 == 0 and X / (C0 / C1) doesn't create UB.
```
Counterexample:
X = INT_MIN, C0 = 2, C1 = -2
C0 / C1 = -1 doesn't create UB, but X / (C0 / C1) does.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 3]
```diff
@@ -365,6 +365,28 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
...
+    const APInt *C1;
+    auto UDivCheck = [&C1](const APInt &C) { return C.urem(*C1).isZero(); };
+    auto SDivCheck = [&C1](const APInt &C) {
+      APInt Quot, Rem;
+      APInt::sdivrem(C, *C1, Quot, Rem);
+      return Rem.isZero() && !Quot.isMinSignedValue();
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:377]:
```suggestion
      return Rem.isZero() && !Quot.isAllOnes();
```
See the above comments.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 4]
```diff
@@ -365,6 +365,28 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
...
+    };
+    if (match(Op1, m_APInt(C1)) &&
+        (match(Op0, m_Exact(m_UDiv(m_Value(X), m_CheckedInt(UDivCheck)))) ||
+         match(Op0, m_Exact(m_SDiv(m_Value(X), m_CheckedInt(SDivCheck)))))) {
+      auto BOpc = cast<BinaryOperator>(Op0)->getOpcode();
+      return BinaryOperator::CreateExact(
+          BOpc, X,
+          Builder.CreateBinOp(BOpc, cast<BinaryOperator>(Op0)->getOperand(1),
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:385]:
```suggestion
          Builder.CreateBinOp(BOpc, RHSC,
```

Can we use `m_CheckedInt(const Constant *&V, ... CheckFn)` ?

File: llvm/test/Transforms/InstCombine/exact.ll [hunk 5]
```diff
@@ -336,3 +336,57 @@ define <2 x i1> @sdiv_icmp6_vec(<2 x i64> %X) {
...
+;
+  %div = udiv exact i8 %x, 11
+  %mul = mul i8 %div, 6
+  ret i8 %mul
+}
+
+define i8 @mul_of_sdiv_fail_ub(i8 %x) {
+; CHECK-LABEL: @mul_of_sdiv_fail_ub(
+; CHECK-NEXT:    [[X_FR:%.*]] = freeze i8 [[X:%.*]]
+; CHECK-NEXT:    [[TMP1:%.*]] = srem i8 [[X_FR]], 6
```
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/exact.ll:385]: I don't think this test case covers UB handling logic.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]
IR diff LGTM.

File: llvm/test/Transforms/InstCombine/exact.ll (hunk 5 above)
Comment (goldsteinn) [reply llvm/test/Transforms/InstCombine/exact.ll:385]: Yeah, needs exact.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp (hunk 4 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:385]: B.c `CreateBinOp` needs non-constant `Value *` but `m_CheckedInt` sets a `const Constant *`. Im a but reluctant to use `const_cast`. Although maybe in a follow up could had non-const pointer support in `cstval_pred_ty`.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!

File: llvm/test/Transforms/InstCombine/exact.ll [hunk 6]
```diff
@@ -336,3 +336,55 @@ define <2 x i1> @sdiv_icmp6_vec(<2 x i64> %X) {
...
+
+define i8 @mul_of_sdiv_fail_ub(i8 %x) {
+; CHECK-LABEL: @mul_of_sdiv_fail_ub(
+; CHECK-NEXT:    [[MUL:%.*]] = sub i8 0, [[X:%.*]]
+; CHECK-NEXT:    ret i8 [[MUL]]
+;
+  %div = sdiv exact i8 %x, 6
+  %mul = mul i8 %div, -6
+  ret i8 %mul
+}
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/exact.ll:390]: As you went out of your way to add non-splat vector support, can you please also add a test for it?

**Review by nikic — APPROVED** [review]
LGTM

**Pull request closed without merging**